File: CSharpRewrite/CSharpRewrite/Assignments/Musicians/BandSupervisor.cs
using System.Threading.Channels;

namespace CSharpRewrite.Assignments.Musicians;

public static class BandSupervisor
{
    public static async Task StartBand(int maxRetries)
    {
        var channel = Channel.CreateUnbounded<(string, SkillLevel)>();
        var senders = new Dictionary<string, ChannelWriter<int>>
        {
            { "singer", AddBandMember("singer", SkillLevel.Good, channel.Writer) },
            { "bass", AddBandMember("bass", SkillLevel.Good, channel.Writer) },
            { "drum", AddBandMember("drum", SkillLevel.Bad, channel.Writer) },
            { "guitar", AddBandMember("guitar", SkillLevel.Good, channel.Writer) },
        };

        while (true)
        {
            var (role, skillLevel) = await channel.Reader.ReadAsync();
            maxRetries--;
            senders.Remove(role);
            if (maxRetries <= 0)
            {
                Console.WriteLine("The manager is mad and fired the whole band!");
                foreach (var (_, writer) in senders)
                {
                    await writer.WriteAsync(0);
                }

                await Task.Delay(TimeSpan.FromSeconds(3));
                break;
            }
            
            senders.Add(role, AddBandMember(role, skillLevel, channel.Writer));
        }
    }

    private static ChannelWriter<int> AddBandMember(string role, SkillLevel skillLevel, ChannelWriter<(string, SkillLevel)> writer)
    {
        var fireChannel = Channel.CreateUnbounded<int>();
        _ = Task.Run(async () =>
        {
            var musician = new Musician(role, skillLevel, fireChannel.Reader);
            while (true)
            {
                await Task.Delay(TimeSpan.FromMilliseconds(Musician.Delay));
                var playResult = musician.PlaySound();
                if (playResult) continue;

                await writer.WriteAsync((role, skillLevel));
                break;
            }
        });

        return fireChannel.Writer;
    }
}

File: KotlinRewrite/src/main/kotlin/io/github/deadshot465/assignments/musicians/bandSupervisor.kt
package io.github.deadshot465.assignments.musicians

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

suspend fun startBand(maxRetries: Int) {
    val channel = Channel<Pair<String, SkillLevel>>(10)
    val members = mutableMapOf(
        "singer" to addBandMember("singer", SkillLevel.Good, channel),
        "bass" to addBandMember("bass", SkillLevel.Good, channel),
        "drum" to addBandMember("drum", SkillLevel.Bad, channel),
        "guitar" to addBandMember("guitar", SkillLevel.Good, channel)
    )

    var retries = maxRetries
    while (true) {
        val data = channel.receive()
        retries -= 1
        members.remove(data.first)

        if (retries <= 0) {
            println("The manager is mad and fired the whole band!")
            for (chan in members.values) {
                chan.send(0)
            }
            delay(3000)
            break
        }

        members[data.first] = addBandMember(data.first, data.second, channel)
    }
}

private fun addBandMember(role: String, skillLevel: SkillLevel, channel: Channel<Pair<String, SkillLevel>>): Channel<Int> {
    val fireChannel = Channel<Int>(10)
    val musician = Musician(role, skillLevel, fireChannel)
    CoroutineScope(Dispatchers.Default).launch {
        while (true) {
            delay(delay)
            val playResult = musician.playSound()
            if (playResult) {
                continue
            }

            channel.send(Pair(role, skillLevel))
            break
        }
    }

    return fireChannel
}